### ISSUE TITLE
Segmentation fault on handling SIGTERM due to race in destruction and usage of static objects

### ISSUE THREAD

**Issue by elinor-fung** [dotnet/core-setup#7913]
## Steps to reproduce
- Start an ASP.NET Core application
- Send `SIGTERM` to the process

## Expected behavior
Application shuts down properly

## Actual behavior
Intermittent segfault in the host due to a race:
- Thread 2 receives and handles the `SIGTERM` signal, which destructs static objects
  - One of these static objects is `hostpolicy_context_t`, which holds on to a `coreclr_t` (in libhostpolicy)
- Thread 1 finishes executing the managed assembly and then calls `coreclr_t::shutdown`
- However, thread 2 has destructed `hostpolicy_context_t`, so the `coreclr_t` instance is no longer valid, resulting in the segfault when `coreclr_t::shutdown` tries to lock the mutex (member on `coreclr_t`)

```
thread #1: tid = 50051, 0x00007f51e67e2fa0 libpthread.so.0`__GI___pthread_mutex_lock, name = 'dotnet', stop reason = signal SIGSEGV
    frame #0: 0x00007f51e67e2fa0 libpthread.so.0`__GI___pthread_mutex_lock (mutex=0x0000000000000000) at pthread_mutex_lock.c:65
    frame #1: 0x00007f51e540512c libhostpolicy.so`coreclr_t::shutdown(int*) [inlined] __gthread_mutex_lock(pthread_mutex_t*) + 28 at gthr-default.h:748
    frame #2: 0x00007f51e540511a libhostpolicy.so`coreclr_t::shutdown(int*) [inlined] std::mutex::lock() at mutex:134
    frame #3: 0x00007f51e540511a libhostpolicy.so`coreclr_t::shutdown(int*) [inlined] std::lock_guard<std::mutex>::lock_guard(std::mutex&) at mutex:414
    frame #4: 0x00007f51e540511a libhostpolicy.so`coreclr_t::shutdown(this=0x0000000000000000, latchedExitCode=0x00007ffd85757b90) + 10 at coreclr.cpp:175
    frame #5: 0x00007f51e54101ec libhostpolicy.so`run_app_for_context(context=<unavailable>, argc=<unavailable>, argv=<unavailable>) + 796 at hostpolicy.cpp:255
    frame #6: 0x00007f51e5410a1c libhostpolicy.so`::corehost_main(argc=6, argv=<unavailable>) + 204 at hostpolicy.cpp:389
    <...>

thread #2: tid = 50067, 0x00007f51e593c991 libc.so.6`__GI___libc_free + 4 at malloc.c:3101, stop reason = signal SIGSEGV
    frame #0: 0x00007f51e593c991 libc.so.6`__GI___libc_free + 4 at malloc.c:3101
    frame #1: 0x00007f51e593c98d libc.so.6`__GI___libc_free(mem=0x0000000000895040) + 61
    frame #2: 0x00007f51e540d629 libhostpolicy.so`deps_entry_t::~deps_entry_t() [inlined] std::string::_Rep::_M_dispose(std::allocator<char> const&) + 24 at basic_string.h:249
    <...>
    frame #20: 0x00007f51e54129a4 libhostpolicy.so`hostpolicy_init_t::~hostpolicy_init_t(this=<unavailable>) + 148 at hostpolicy_init.h:12
    frame #21: 0x00007f51e58e8041 libc.so.6`__run_exit_handlers(status=0, listp=0x00007f51e5c90718, run_list_atexit=<unavailable>, run_dtors=<unavailable>) + 369 at exit.c:108
    frame #22: 0x00007f51e58e813a libc.so.6`__GI_exit(status=<unavailable>) + 26 at exit.c:139
    frame #23: 0x00007f51e51107af libcoreclr.so`PROCEndProcess(hProcess=<unavailable>, uExitCode=0, bTerminateUnconditionally=NO) + 127 at process.cpp:1484
    frame #24: 0x00007f51e511071f libcoreclr.so`::ExitProcess(uExitCode=<unavailable>) + 143 at process.cpp:1335
    frame #25: 0x00007f51e4e2af79 libcoreclr.so`EEPolicy::ExitProcessViaShim(exitCode=<unavailable>) + 9 at eepolicy.cpp:538
    frame #26: 0x00007f51e4e2af3a libcoreclr.so`SafeExitProcess(exitCode=0, fAbort=NO, sca=SCA_ExitProcessWhenShutdownComplete) + 490 at eepolicy.cpp:515
    frame #27: 0x00007f51e510b1e7 libcoreclr.so`CorUnix::TerminationRequestHandlingRoutine(pArg=<unavailable>) + 23 at synchmanager.cpp:1666
    frame #28: 0x00007f51e51137ad libcoreclr.so`CorUnix::CPalThread::ThreadEntry(pvParam=0x00007f51d8000bb0) + 349 at thread.cpp:1807
    frame #29: 0x00007f51e67e06db libpthread.so.0`start_thread + 219
    frame #30: 0x00007f51e59c688f libc.so.6`__GI___clone + 63 at clone.S:95
```

**Comment by elinor-fung** [comment]
Test in aspnetcore hitting this:
https://github.com/aspnet/AspNetCore/blob/master/src/Hosting/test/FunctionalTests/ShutdownTests.cs

**Comment by elinor-fung** [comment]
Merged into release/3.1 with #8290
